### PR TITLE
id24: skies

### PR DESCRIPTION
**Pull request by rfomin**
Legacy of Rust, map 07:

https://github.com/user-attachments/assets/6346364f-4d72-4edb-b47e-f6cbc3057235

### CODE REVIEW

File: src/r_plane.c [hunk 1]
```diff
@@ -376,152 +382,342 @@ static void R_MakeSpans(int x, unsigned int t1, unsigned int b1, unsigned int t2
     spanstart[b2--] = x;
 }
 
-// New function, by Lee Killough
+#define FIRE_WIDTH     64
+#define FIRE_HEIGHT    256
 
-static void do_draw_plane(visplane_t *pl)
+static byte fire_indexes[FIRE_WIDTH * FIRE_HEIGHT];
```
Comment (fabiangreffrath) [inline src/r_plane.c:388]: I think the correct term is `indices`.

File: src/r_plane.c [hunk 2]
```diff
@@ -376,152 +382,342 @@ static void R_MakeSpans(int x, unsigned int t1, unsigned int b1, unsigned int t2
...
+
+static byte fire_pixels[FIRE_WIDTH * FIRE_HEIGHT];
+
+void PrepareFirePixels(fire_t *fire)
```
Comment (fabiangreffrath) [inline src/r_plane.c:392]: static

File: src/r_plane.c [hunk 3]
```diff
@@ -376,152 +382,342 @@ static void R_MakeSpans(int x, unsigned int t1, unsigned int b1, unsigned int t2
...
+    memset(fire_indexes, 0, FIRE_WIDTH * FIRE_HEIGHT);
+
+    int last = array_size(fire->palette) - 1;
+
+    for (int i = 0; i < FIRE_WIDTH; ++i)
+    {
+        fire_indexes[(FIRE_HEIGHT - 1) * FIRE_WIDTH + i] = last;
+    }
+
+    for (int i = 0; i < 64; ++i)
```
Comment (fabiangreffrath) [inline src/r_plane.c:441]: 64 is FIRE_WIDTH?

File: src/r_plane.c [hunk 4]
```diff
@@ -376,152 +382,342 @@ static void R_MakeSpans(int x, unsigned int t1, unsigned int b1, unsigned int t2
...
+        {
+            *rover++ = fire->palette[*src];
+            src += FIRE_WIDTH;
+        }
+    }
+}
+
+static void SpreadFire(void)
+{
+    for(int x = 0; x < FIRE_WIDTH; ++x)
```
Comment (fabiangreffrath) [inline src/r_plane.c:408]: missing space

File: src/r_plane.c (hunk 1 above)
Comment (rfomin) [reply src/r_plane.c]: Fixed in https://github.com/fabiangreffrath/woof/pull/1894/commits/55aa873a86972b39a5fd1e962a27615089c9441b

File: src/r_plane.c (hunk 2 above)
Comment (rfomin) [reply src/r_plane.c]: Fixed in https://github.com/fabiangreffrath/woof/pull/1894/commits/55aa873a86972b39a5fd1e962a27615089c9441b

File: src/r_plane.c (hunk 3 above)
Comment (rfomin) [reply src/r_plane.c:441]: This is a requirement of the spec.
> Fire skies are expected to be primed with both the bottom row set to the last palette index; and with a minimum of 64 iterations performed to populate the target image with some fire to begin with.

https://docs.google.com/document/d/1RWNjk3mr3CLFptGCSVkZl5hFEPr1R5OzeSu7aHY8Zwg/edit?usp=sharing

File: src/r_plane.c [hunk 5]
```diff
@@ -376,152 +382,342 @@ static void R_MakeSpans(int x, unsigned int t1, unsigned int b1, unsigned int t2
     spanstart[b2--] = x;
 }
 
-// New function, by Lee Killough
+#define FIRE_WIDTH     64
+#define FIRE_HEIGHT    256
```
[the remark this answers is not in the document]
Comment (rfomin) [reply src/r_plane.c:386]: Tweaked here: https://github.com/fabiangreffrath/woof/pull/1894/commits/55aa873a86972b39a5fd1e962a27615089c9441b

On the other hand, LoR MAP07 now looks boring. 😄

File: src/r_plane.c (hunk 4 above)
Comment (rfomin) [reply src/r_plane.c]: Fixed here https://github.com/fabiangreffrath/woof/pull/1894/commits/0314b3e25196036842747820c6eb5b20b1caa037

File: src/r_plane.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (fabiangreffrath) [reply src/r_plane.c:386]: I wouldn't call it like that. Any change to get the tips of the flames a bit higher, though?

[the remark this answers is not in the document]
Comment (rfomin) [reply src/r_plane.c:386]: > Any change to get the tips of the flames a bit higher, though?

I don't know how to do this. I think it currently works just like the KEX port (and the original PSX). Let's leave it as it is for now.

Here's a full description of the algorithm: https://fabiensanglard.net/doom_fire_psx/

[the remark this answers is not in the document]
Comment (fabiangreffrath) [reply src/r_plane.c:386]: > I don't know how to do this. I think it currently works just like the KEX port (and the original PSX). Let's leave it as it is for now.

Alright, agreed.

> Here's a full description of the algorithm: https://fabiensanglard.net/doom_fire_psx/

Yes, I've seen that. A nice ressource of knowledge!

File: src/r_plane.c [hunk 6]
```diff
@@ -376,152 +382,343 @@ static void R_MakeSpans(int x, unsigned int t1, unsigned int b1, unsigned int t2
...
+        byte *src = fire_indices + x;
+        for (int y = 0; y < FIRE_HEIGHT; y++)
+        {
+            *rover++ = fire->palette[*src];
+            src += FIRE_WIDTH;
+        }
+    }
+}
+
+static void SpreadFire(void)
```
Comment (fabiangreffrath) [inline src/r_plane.c:406]: Does it make sense to factor out the fire stuff into its own source file?

Comment (rfomin) [reply src/r_plane.c:406]: I don't think it's worth it, it's very small. We can move most of it to r_sky.c or r_skydef.c.

Comment (fabiangreffrath) [reply src/r_plane.c:406]: Okay.

Comment (rfomin) [reply src/r_plane.c]: Fixed in https://github.com/fabiangreffrath/woof/pull/1894/commits/ee75c3a915d119c4f3dfda1f40aabb1a4a57cea3

File: src/r_plane.c [hunk 7]
```diff
@@ -376,152 +383,258 @@ static void R_MakeSpans(int x, unsigned int t1, unsigned int b1, unsigned int t2
...
+
+    dc_colormap[0] = dc_colormap[1] = fullcolormap;
+
+    dc_texturemid = skytex->mid * FRACUNIT;
+    dc_texheight = textureheight[texture] >> FRACBITS;
+    dc_iscale = FixedMul(skyiscale, skytex->scaley);
+
+    dc_texturemid += skytex->curry;
+
+    angle_t an = viewangle + FixedToAngle(skytex->currx);
```
Comment (rfomin) [inline src/r_plane.c:421]: Scrolling/scaling works, but not exactly as in the KEX port. R&R implementation is complicated: https://github.com/GooberMan/rum-and-raisin-doom/blob/4af881489bda0b26f12744de05dba7a571f2faa9/src/doom/r_sky.cpp#L429

Test WAD: [sky.zip](https://github.com/user-attachments/files/16982403/sky.zip)

File: src/r_plane.c [hunk 8]
```diff
@@ -449,7 +450,8 @@ static void DrawSkyDef(visplane_t *pl)
 
     if (sky->type == SkyType_WithForeground)
     {
-        colfunc = R_DrawSkyColumnMasked;
+        tranmap = W_CacheLumpName("SKYTRAN", PU_CACHE);
```
Comment (fabiangreffrath) [inline src/r_plane.c:446]: He, what is this?

Comment (rfomin) [reply src/r_plane.c:446]: This is the trick that allows to use `R_DrawTLColumn` for drawing foreground sky. Quote from the spec:
> A custom render path is not required for the foreground sky to render transparently. All that
is needed is a correctly cooked transparency map, which results in the backbuffer texel
being used when the texture texel is a 0 index. While this map is generated at compile time
for our implementation, a copy of its contents can be found as SKYTRAN.lmp in the bundled
resources.

**Review by fabiangreffrath — APPROVED** [review]
Great job, thank you!

File: src/r_plane.c (hunk 8 above)
Comment (fabiangreffrath) [reply src/r_plane.c:446]: Maybe you could leave a comment about this? I'm not sure if I'll still remember this next year. :wink:

Comment (rfomin) [reply src/r_plane.c:448]: Fixed in https://github.com/fabiangreffrath/woof/pull/1894/commits/80855cc3114e32d6a1d4f7772327e5265eab83a2

**Pull request merged**